File: srdashboard/srql.go
package srdashboard

import (
	"sort"
	"strconv"
	"strings"
)

// SRQLQuery describes a simple key:value SRQL query for dashboard interactions.
type SRQLQuery struct {
	Entity      string
	SearchField string
	Search      string
	Include     map[string][]string
	Exclude     map[string][]string
	Where       []string
	Limit       int
}

// BuildSRQL builds a deterministic SRQL query string for dashboard packages.
func BuildSRQL(query SRQLQuery) string {
	entity := strings.TrimSpace(query.Entity)
	if entity == "" {
		entity = "devices"
	}

	tokens := []string{"in:" + entity}

	if search := strings.TrimSpace(query.Search); search != "" && strings.TrimSpace(query.SearchField) != "" {
		tokens = append(tokens, strings.TrimSpace(query.SearchField)+":%"+EscapeSRQLValue(search)+"%")
	}

	tokens = appendFilters(tokens, query.Include, "")
	tokens = appendFilters(tokens, query.Exclude, "!")

	for _, clause := range query.Where {
		if text := strings.TrimSpace(clause); text != "" {
			tokens = append(tokens, text)
		}
	}

	if query.Limit > 0 {
		tokens = append(tokens, "limit:"+strconv.Itoa(query.Limit))
	}

	return strings.Join(tokens, " ")
}

// EscapeSRQLValue escapes whitespace in a single SRQL value token.
func EscapeSRQLValue(value string) string {
	return strings.Join(strings.Fields(strings.TrimSpace(value)), `\ `)
}

// SRQLList builds a parenthesized SRQL list from values.
func SRQLList(values []string) string {
	escaped := make([]string, 0, len(values))
	for _, value := range values {
		if text := EscapeSRQLValue(value); text != "" {
			escaped = append(escaped, text)
		}
	}

	return "(" + strings.Join(escaped, ",") + ")"
}

func appendFilters(tokens []string, filters map[string][]string, prefix string) []string {
	keys := make([]string, 0, len(filters))
	for key := range filters {
		if strings.TrimSpace(key) != "" {
			keys = append(keys, key)
		}
	}
	sort.Strings(keys)

	for _, key := range keys {
		values := compactValues(filters[key])
		if len(values) == 0 {
			continue
		}

		tokens = append(tokens, prefix+strings.TrimSpace(key)+":"+SRQLList(values))
	}

	return tokens
}

func compactValues(values []string) []string {
	out := make([]string, 0, len(values))
	for _, value := range values {
		if text := strings.TrimSpace(value); text != "" {
			out = append(out, text)
		}
	}
	sort.Strings(out)

	return out
}

File: src/srql.js
export function escapeSrqlValue(value) {
  return String(value ?? "").trim().replace(/\s+/g, "\\ ")
}

export function srqlList(values) {
  return `(${Array.from(values || []).map(escapeSrqlValue).filter(Boolean).join(",")})`
}

export function buildSrqlQuery(options = {}) {
  const entity = String(options.entity || "devices").trim() || "devices"
  const tokens = [`in:${entity}`]
  const search = String(options.search || "").trim()
  const searchField = String(options.searchField || "").trim()

  if (search && searchField) tokens.push(`${searchField}:%${escapeSrqlValue(search)}%`)
  appendFilters(tokens, options.include, "")
  appendFilters(tokens, options.exclude, "!")

  for (const clause of Array.from(options.where || [])) {
    const text = String(clause || "").trim()
    if (text) tokens.push(text)
  }

  const limit = Number(options.limit)
  if (Number.isInteger(limit) && limit > 0) tokens.push(`limit:${limit}`)

  return tokens.join(" ")
}

export function createSrqlClient(api) {
  const fallback = api?.srql || {}
  const query = typeof fallback.query === "function" ? fallback.query.bind(fallback) : () => ""
  const srqlUpdate = typeof fallback.update === "function" ? fallback.update.bind(fallback) : null
  const legacyUpdate = typeof api?.setSrqlQuery === "function" ? api.setSrqlQuery.bind(api) : null
  const update = (query, frameQueries = {}) => {
    if (srqlUpdate) srqlUpdate(query, frameQueries)
    if (legacyUpdate && fallback.update !== api?.setSrqlQuery) legacyUpdate(query, frameQueries)
  }

  return {
    query,
    update,
    updateQuery: update,
    setQuery: update,
    escapeValue: escapeSrqlValue,
    list: srqlList,
    build: buildSrqlQuery,
  }
}

function appendFilters(tokens, filters, prefix) {
  const keys = Object.keys(filters || {}).filter((key) => String(key || "").trim()).sort()

  for (const key of keys) {
    const values = Array.from(filters[key] || []).filter(Boolean)
    if (values.length > 0) tokens.push(`${prefix}${String(key).trim()}:${srqlList(values)}`)
  }
}
